RoutingService -> MessageService refactor

tplooker opened the pull request:
#### Short description of what this resolves:

This PR is essentially preparation work for adding support for message threading and other more advanced message processing features. Also included in the PR is a slight modification of the handler architecture.

#### Changes proposed in this pull request:

- Refactored IRouterService -> IMessageService
- Refactored DefaultRouterService -> DefaultMessageService
- Extended IMessageService and implementation to include a new RecieveAsync method that returns a message context object.
- Shifted the AgentWireMessage processing logic located in the AgentBase into the RecieveAsync method.
- Modified the handler architecture to allow for greater single point configuration from the top level API, added in new extension methods to demonstrate this with AspNetCore targets.
- Enforced the standard MIME type for agent messages transmitting across HTTP(/s)

**Fixes**: #71 
**Related To**: #64

tmarkovski commented:
You may need to do another PR after merging for the samples project. Since I removed the project references and replaced them with nuget package. Samples should represent a real example of usage, adding the source code doesn't make sense.

tmarkovski commented:
Should we rename the `UseAgent` and `AddAgent` to `UseAgentFramework` and `AddAgentFramework` instead? Seems more in line with the brand :)

tplooker commented on samples/aspnetcore/Messages/PrivateMessageHandler.cs:
```diff
@@ -26,5 +31,20 @@ protected override Task ProcessAsync(PrivateMessage message, ConnectionContext c
+        public IEnumerable<string> SupportedMessageTypes => new[] { PrivateMessageType };
+
+        public async Task ProcessAsync(MessageContext agentMessageContext, ConnectionContext connectionContext)
+        {
```
My thought was to drop the connection property and refactor ConnectionContext -> AgentContext as that is what it is, which is distinctly different to MessageContext? Could be convinced otherwise.

tmarkovski replied:
That's also a good approach. AgentContext closer identifies the problem. MessageContext isn't exactly a context then, it's just an unformatted message payload, which is always a JSON, so we can easily change the input type to JObject, which is an upgrade from a string or byte array, but can easily be converted to typed object by handlers using `.ToObject<T>()`

tplooker replied:
But as we add message threading support I would counter that it will become more of a context object?

tmarkovski replied:
Good point. I guess it just looks too complex having all these contexts as parameters. Perhaps move the AgentContext in the MessageContext? That may make more sense, since the message does belong a Connection channel, which is tied to an Agent which has wallet and pool?

tplooker replied:
Yeah sounds good, will update accordingly!

tplooker commented on src/AgentFramework.AspNetCore/Configuration/Service/ServiceBuilderExtensions.cs:
```diff
@@ -11,25 +12,41 @@ namespace AgentFramework.AspNetCore.Configuration.Service
+            return builder;
+        }
 
+        public static AgentServicesBuilder OverrideCoreMessageHandlers(this AgentServicesBuilder builder)
```
We have used Add for extension methods and that seems the pattern in .AspNetCore for adding services to the DI pool? How about AddDefaultMessageHandlers, I didnt want the user to have to supply this in the top level API when configuring it out of the box hence the .OverrideCoreMessageHandlers() method? I will rename it to OverrideDefaultMessageHandlers.

tplooker commented:
> You may need to do another PR after merging for the samples project. Since I removed the project references and replaced them with nuget package. Samples should represent a real example of usage, adding the source code doesn't make sense.

> I'll just leave this comment here, since you're focused on the messaging part. Moving forward, I'd like us to think about the MessageService working directly with `IAgentMessage` messages (or wrapped in WireMessage), meaning it is only responsible for processing them using Auth Pack/Unpack. It then hands them over to the some driver or endpoint service `IEndpointService` which is implemented by `HttpEndpointService` that accepts an `AgentEndpoint` or `EndpointDescription` which will wrap the agent message into a `ForwardMessage` or whatever format agreed upon and do the final AnonPack/Unpack. This is just a note and point of discussion, doesn't need to go with this PR.

I Agree with this conceptually, not entirely sure how we will draw that abstraction yet.

tplooker commented on src/AgentFramework.Core.Handlers/MessageHandlerBase.cs:
```diff
@@ -1,41 +0,0 @@
-﻿using System.Collections.Generic;
```
Im not against it can add back in there was just no example usage, any recommendation on another name as MessageHandlerBase implies to me all MessageHandlers derived from it, SingleMessageHandlerBase?

tmarkovski replied:
The `PrivateMessageHandler` was using it, specifying handler per message pattern. I think the name MessageHandlerBase (or AgentMessageHandlerBase) is fine. You may be getting too descriptive trying to capture entire functionality in the class name. We'll have documentation and code comments to reflect more of what the usage and intention of this is.

tplooker replied:
Cool, yeap just re-adding now

tmarkovski commented:
> I Agree with this conceptually, not entirely sure how we will draw that abstraction yet.

I think I was just thinking out loud here. I'll share some thoughts on this soon.

tplooker commented on src/AgentFramework.AspNetCore/Configuration/Service/ServiceBuilderExtensions.cs:
```diff
@@ -11,25 +12,41 @@ namespace AgentFramework.AspNetCore.Configuration.Service
     {
-        public static AgentServicesBuilder AddCoreServices(this AgentServicesBuilder builder)
+        internal static AgentServicesBuilder AddCoreServices(this AgentServicesBuilder builder)
         {
```
Yeap good catch, I agree.
